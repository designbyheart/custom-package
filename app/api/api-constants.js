// @flow

export const TOKEN_EXPIRED_CODE = 'MS-107'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_SENDER_DID = 'sD'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'sVk'
export const QR_CODE_LOGO_URL = 'lu'
export const QR_CODE_REQUEST_ID = 'rid'
export const QR_CODE_AGENT_PROOF = 'sakdp'
export const QR_CODE_SENDER_NAME = 'sn'
export const QR_CODE_TARGET_NAME = 'tn'
export const QR_CODE_SENDER_ENDPOINT = 'e'

export const MESSAGE_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'claimOffer',
  CLAIM_REQUEST: 'claimReq',
  PROOF_REQUEST: 'proofReq',
  CLAIM: 'claim',
}

export const API_TYPE = {
  CONNECT: 'CONNECT',
  REGISTER: 'REGISTER',
  CREATE_AGENT: 'CREATE_AGENT',
  INVITE_ANSWERED: 'INVITE_ANSWERED',
  CREATE_KEY: 'CREATE_KEY',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
  SEND_MESSAGE: 'SEND_MSG',
}

export const STATUS_CODE = {
  ACCEPT: 'MS-104',
}

export const ERROR_ALREADY_EXIST = {
  code: 'OCS',
  message: 'duplicate connection request',
}

export const ERROR_INVITATION_RESPONSE_PARSE_CODE = 'RN-100'

export const ERROR_INVITATION_RESPONSE_PARSE =
  'Error parsing error message in invitation response'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE = 'RN-101'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE =
  'Error parsing error message Pending Invitation SMS'
