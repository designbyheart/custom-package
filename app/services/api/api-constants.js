// @flow
// export const TOKEN_EXPIRED_CODE = 'CRS-121'
export const TOKEN_EXPIRED_CODE = 'MS-107'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_SENDER_DID = 'sD'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'sVk'
export const QR_CODE_LOGO_URL = 'lu'
export const QR_CODE_REQUEST_ID = 'rid'
export const QR_CODE_AGENT_PROOF = 'sakdp'
export const QR_CODE_SENDER_NAME = 'sn'
export const QR_CODE_TARGET_NAME = 'tn'
export const QR_CODE_SENDER_ENDPOINT = 'e'

// TODO: Delete these constants once QR Code is fixed as per new schema
export const QR_CODE_CHALLENGE = 'c'
export const QR_CODE_REMOTE_PAIR_WISE_DID = 'rpDID'
export const QR_CODE_REMOTE_HOSTING_DID = 'rhDID'

export const PUSH_NOTIFICATION_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'claimOffer',
}

export const API_TYPE = {
  CONNECT: 'CONNECT',
  REGISTER: 'REGISTER',
  CREATE_AGENT: 'CREATE_AGENT',
  INVITE_ANSWERED: 'INVITE_ANSWERED',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
}
