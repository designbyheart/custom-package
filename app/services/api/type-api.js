// @flow

export type ApiData = {
  method: string,
  mode: string,
  headers: { [string]: string },
  body?: string,
}

export type BackendError = {
  statusCode: string,
  statusMsg: string,
}
