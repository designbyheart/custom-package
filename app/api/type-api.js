// @flow
import type { GenericObject } from '../common/type-common'
import { MESSAGE_TYPE } from './api-constants'
import type { IndyClaimRequest } from '../bridge/react-native-cxs/type-cxs'

export type ApiData = {
  method: string,
  mode: string,
  headers: { [string]: string },
  body?: string,
}

export type BackendError = {
  statusCode: string,
  statusMsg: string,
}

export type GetInvitationLinkApiData = {
  agencyUrl: string,
  smsToken: string,
}

export type MessageApiData = {
  agencyUrl: string,
  dataBody: GenericObject,
}

export type DownloadInvitationApiData = {
  url: string,
}

export type SendAuthenticationResponseApiData = {
  data: {
    identifier: string,
    dataBody: GenericObject,
  },
  config: {
    agencyUrl: string,
  },
}

export type GetProfileApiData = {
  identifier: string,
  challenge: string,
  signature: string,
  agencyUrl: string,
}

export type SendMessageApiData = {
  agencyUrl: string,
  userPairwiseDid: string,
  agentPayload: GenericObject,
}

export type ApiClaimRequest = {
  userPairwiseDid: string,
  remoteDid: string,
} & IndyClaimRequest

export type SendClaimRequestApiData = {
  agencyUrl: string,
  userPairwiseDid: string,
  responseMsgId: string,
  claimRequest: ApiClaimRequest,
}

export type EdgeClaimRequest = {
  msg_type: typeof MESSAGE_TYPE.CLAIM_REQUEST,
  version: string,
  to_did: string,
  from_did: string,
  tid: string,
  mid: string,
} & IndyClaimRequest
