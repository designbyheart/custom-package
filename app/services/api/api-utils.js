import { SERVER_ERROR_CODE } from '../../common/api-constants'

export const options = (method = 'GET', body) => {
  let data = {
    method,
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  }
  if (body) {
    data.body = JSON.stringify(body)
  }

  return data
}

// TODO: Write common code for API calling
export const api = (url, apiOptions) =>
  fetch(url, apiOptions)
    .then(res => {
      // TODO:KS create common method to return successful
      // and unsuccessful api response
      if (res.status >= 200 && res.status < 300) {
        return res
          .json()
          .then(response => ({
            payload: response,
          }))
          .catch(error => ({
            // our api call was successful,
            // however server did not return any json response
            // Ideally server should send such response with status code 204
            // so that we don't try to parse response for such methods.
            // As of now just send an empty payload signalling that call went fine
            payload: {},
          }))
      } else {
        // Fail with error code if status code is above 300
        return res.text().then(response => ({
          error: response,
        }))
      }
    })
    .then(response => {
      // if response contains payload and no error, that means we got success
      if (response.payload && !response.error) {
        return response.payload
      } else {
        let errorResponse = {
          statusMsg: 'Server error',
          statusCode: SERVER_ERROR_CODE,
        }

        try {
          // try to convert error response to json, if it fails that means
          // we did not get error code and message
          errorResponse = JSON.parse(response.error)
        } finally {
          // since we did not get error code and message
          // let's just use default that we assigned above,
          // we don't need to do anything here
        }

        throw new Error(JSON.stringify(errorResponse))
      }
    })
