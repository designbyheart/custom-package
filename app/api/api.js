// @flow
import { api, options } from './api-utils'
import type {
  GetInvitationLinkApiData,
  MessageApiData,
  DownloadInvitationApiData,
  SendAuthenticationResponseApiData,
  GetProfileApiData,
  SendMessageApiData,
  SendClaimRequestApiData,
  ApiClaimRequest,
  EdgeClaimRequest,
} from './type-api'
import { PAYLOAD_TYPE, MESSAGE_TYPE, STATUS_CODE } from './api-constants'

export const getInvitationLink = ({
  agencyUrl,
  smsToken,
}: GetInvitationLinkApiData) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }: DownloadInvitationApiData) =>
  api(url, options('GET'))

export const connectWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const registerWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const createAgentWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const createAgentPairwiseKey = ({
  agencyUrl,
  dataBody,
}: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendInvitationResponse = ({
  agencyUrl,
  dataBody,
}: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const getAdditionalData = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}: SendAuthenticationResponseApiData) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const getProfile = ({
  identifier,
  challenge,
  signature,
  agencyUrl,
}: GetProfileApiData) =>
  api(
    `${agencyUrl}/agent/${identifier}/remote/profile?challenge=${challenge}&signature=${signature}`,
    options('GET')
  )

export const sendUpdatedPushToken = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendMessage = ({
  agencyUrl,
  userPairwiseDid,
  agentPayload,
}: SendMessageApiData) =>
  api(
    `${agencyUrl}/agency/route`,
    options('POST', {
      to: userPairwiseDid,
      agentPayload: JSON.stringify(agentPayload),
    })
  )

export function convertClaimRequestToEdgeClaimRequest(
  claimRequest: ApiClaimRequest
): EdgeClaimRequest {
  const { blinded_ms, schema_seq_no, issuer_did } = claimRequest

  return {
    blinded_ms,
    issuer_did,
    schema_seq_no,
    msg_type: MESSAGE_TYPE.CLAIM_REQUEST,
    // hard coded version as of now, update once versioning is implemented
    version: '0.1',
    to_did: claimRequest.remoteDid,
    from_did: claimRequest.userPairwiseDid,
    tid: '1',
    mid: '1',
  }
}

export const sendClaimRequest = ({
  agencyUrl,
  userPairwiseDid,
  responseMsgId,
  claimRequest,
}: SendClaimRequestApiData) =>
  sendMessage({
    agencyUrl,
    userPairwiseDid,
    agentPayload: {
      type: PAYLOAD_TYPE.SEND_MESSAGE,
      msgType: MESSAGE_TYPE.CLAIM_REQUEST,
      edgeAgentPayload: JSON.stringify(
        convertClaimRequestToEdgeClaimRequest(claimRequest)
      ),
      refMsgId: responseMsgId,
      // just accept for now, we are not sending rejection as of now
      statusCode: STATUS_CODE.ACCEPT,
    },
  })
