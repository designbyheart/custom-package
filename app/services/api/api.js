import { api, options } from './api-utils'

export const getInvitationLink = ({ agencyUrl, smsToken }) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }) => api(url, options('GET'))

export const connectWithConsumerAgency = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const registerWithConsumerAgency = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const createAgentWithConsumerAgency = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const sendInvitationResponse = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const getAdditionalData = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const getProfile = ({ identifier, challenge, signature, agencyUrl }) =>
  api(
    `${agencyUrl}/agent/${identifier}/remote/profile?challenge=${challenge}&signature=${signature}`
  )

export const sendQRInvitationResponse = ({ challenge, signature, agencyUrl }) =>
  api(
    `${agencyUrl}/agent/connection-req`,
    options('POST', { challenge, signature }),
    true
  )

export const sendUpdatedPushToken = ({ agencyUrl, dataBody }) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))
