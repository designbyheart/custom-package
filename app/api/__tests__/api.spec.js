// @flow
import renderer from 'react-test-renderer'
import { convertClaimRequestToEdgeClaimRequest } from '../api'

describe('convertClaimRequestToEdgeClaimRequest', () => {
  it('should convert correctly', () => {
    const edgeClaimRequest = convertClaimRequestToEdgeClaimRequest({
      blinded_ms: {
        prover_did: 'userPairwiseDid',
        u: 'u',
        ur: 'ur',
      },
      schema_seq_no: 12,
      issuer_did: 'issuer_did',
      remoteDid: 'remoteDid',
      userPairwiseDid: 'userPairwiseDid',
    })
    expect(edgeClaimRequest).toMatchSnapshot()
  })
})
