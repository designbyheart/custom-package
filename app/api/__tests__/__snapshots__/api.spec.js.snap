// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertClaimRequestToEdgeClaimRequest should convert correctly 1`] = `
Object {
  "blinded_ms": Object {
    "prover_did": "userPairwiseDid",
    "u": "u",
    "ur": "ur",
  },
  "from_did": "userPairwiseDid",
  "issuer_did": "issuer_did",
  "mid": "1",
  "msg_type": "claimReq",
  "schema_seq_no": 12,
  "tid": "1",
  "to_did": "remoteDid",
  "version": "0.1",
}
`;
